### ISSUE TITLE
adding multiple identical services to an accessory

### ISSUE THREAD

**Issue by snowdd1** [nfarina/homebridge#139]
I tried to add two `Service.Lightbulb()` to one device
HAP complained about missing subtype (ok, was in a PR some time ago).
But I found 
  -  the subtype cannot be used when using the generated types, as the parameter is not passed
  -  when adding the subtypes directly `myService.subtype="string1"` etc. there appear as much services in legato EVE, but they are all the same (the latest added service).
Broken or did just wreck it myself?

**Comment by snowdd1** [comment]
Nice effect however: the Characteristic.Name is shown in EVE if multiple services with same UUID are present. However, as stated above, it is always the latest added, repeated.

**Comment by nfarina** [comment]
You should be able to add multiple Services of the same "type" (like Lightbulb) as long as you specify different display names and subtypes in the constructor - i.e. `new Service.Lightbulb("Display Backlight", "backlight")`

**Comment by snowdd1** [comment]
No, that fails.
The signature in HomeKitTypes.js is
````javascript
Service.Lightbulb = function(displayName) {
````
there is no subtype passed to the super constructor.
````javascript
  Service.call(this, displayName, '00000043-0000-1000-8000-0026BB765291');
````
JUST AS I WRITE my HAP-nodeJS clone is outdated.
Closing.

**Comment by snowdd1** [comment]
`npm update` helped to get the right signature for the constructor. Sorry for the confusion. I had updated homebridge, but not the dependencies. Something I have still to get acquainted to.

**Comment by snowdd1** [comment]
Sorry, changed the behavior but didn't work for me.
I have tried to add   
` var light1 = new Service.Lightbulb("Main","main");`  and
` var light2 = new Service.Lightbulb("Second","second")`
but it shows *Second* twice in EVE, and the events of both in display belong to the later added (i.e. returned in the platform shim) with `return [informationService, light1, light2]`

**Comment by snowdd1** [comment]
I probably have to re-investigate all this, as it seems yesterday at about 0h UTC iCloud was down or at least very slow and did not always send correct data. Another thing I wrote yesterday (knx thermostat shim) didn't work in the night but does this morning (at 8 UTC).
I'll get back to this.

**Comment by snowdd1** [comment]
Even with iCloud being back online and a complete refactoring of my code, adding two services of the same type lead to having the second service twice.

**Comment by snowdd1** [comment]
Another similar issue: when I add multiple *distinct!* services to one device, and try to add homebridge to homekit, it "unpairs" itself after the PIN
````shell
  Accessory [Homebridge] Paired with client 9270ACEE-7894-4197-90B5-61EB96D321CD +34ms
  EventedHTTPServer [192.168.178.84] HTTP Repsonse is finished +23ms
  EventedHTTPServer [192.168.178.84] Client connection closed +17ms
  EventedHTTPServer [192.168.178.84] HTTP connection was closed +2ms
  EventedHTTPServer [192.168.178.84] HTTP server was closed +1ms
  EventedHTTPServer [192.168.178.84] New connection from client +10ms
  EventedHTTPServer [192.168.178.84] HTTP server listening on port 54826 +1ms
  EventedHTTPServer [192.168.178.84] HTTP request: /pair-verify +5ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: POST /pair-verify +1ms
  HAPServer [CC:22:3D:E3:CE:30] Pair verify step 1/2 +2ms
  EventedHTTPServer [192.168.178.84] HTTP Repsonse is finished +21ms
  EventedHTTPServer [192.168.178.84] HTTP request: /pair-verify +19ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: POST /pair-verify +1ms
  HAPServer [CC:22:3D:E3:CE:30] Pair verify step 2/2 +2ms
  HAPServer [CC:22:3D:E3:CE:30] Client 9270ACEE-7894-4197-90B5-61EB96D321CD verification complete +7ms
  EventedHTTPServer [192.168.178.84] HTTP Repsonse is finished +3ms
  EventedHTTPServer [192.168.178.84] HTTP request: /accessories +5ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: GET /accessories +1ms
  EventedHTTPServer [192.168.178.84] HTTP Repsonse is finished +8ms
  EventedHTTPServer [192.168.178.84] HTTP request: /pairings +108ms
  HAPServer [CC:22:3D:E3:CE:30] HAP Request: POST /pairings +1ms
  HAPServer [CC:22:3D:E3:CE:30] Removing pairing info for client +2ms
  Accessory [Homebridge] Unpairing with client 9270ACEE-7894-4197-90B5-61EB96D321CD +1ms
````

**Comment by snowdd1** [comment]
Is there a way to identify what HomeKit is complaining about?

**Comment by nfarina** [comment]
So HomeKit will "unpair" if it doesn't like something about the returned accessories/services.

**Comment by snowdd1** [comment]
Does it remember that it doesn't like a device/bridge? Because when a add a thermostat service (to a new accessory or to an existing accessory) it stops working, if I remove it again to the status before, it doesn't resume talking to homebridge.
I can only reset HomeKit data and `rm persist/* ` to start it again.

**Comment by nfarina** [comment]
So there seems to be a "blocked" state where HomeKit will (temporarily?) blacklist a device that it doesn't like. In MyTouchHome, you can press and hold on the bridge and "Unblock" it. I'm not sure how you do that in the other HomeKit apps though.

**Comment by snowdd1** [comment]
Cool, I have myTouchHome installed though I didn't use any more, EVE has the more user friendly UI.

Strange thing I found though. 
If I add a thermostat accessory with a blank service, i.e. `var myService = new Service.Thermostat()` I can connect to homebridge. If a add listeners to the `CurrentTemperature` characteristic, everything is still fine. But if I do exactly the same to `TargetTemperature` characteristic, HomeKit immediately refuses to talk to me.
I use the same method to bind the characteristic to the events:
https://github.com/snowdd1/homebridge/blob/knx-dev/accessories/knxdevice.js#L373 
I am not adding any more characteristics, just `get` ting the existing and adding events with characteristic.on, and passing the reference to my array of callbacks.
For now I am clueless.

**Comment by KhaosT** [comment]
@snowdd1 on the computer you host the Homebridge, can you try open http://127.0.0.1:51826/accessories and post the JSON here?

**Comment by snowdd1** [comment]
@KhaosT I put both versions as Gists.
Without events added : https://gist.github.com/snowdd1/7710b2dba33d7cb4b284
With events : https://gist.github.com/snowdd1/79ecec24dd1345ce1a56
I think they are identical. I am lost :frowning:

**Comment by snowdd1** [comment]
> In MyTouchHome, you can press and hold on the bridge and "Unblock" it.

@nfarina How do you do that? If I hold the bridge on my iPad it doesn't do anything.

**Comment by snowdd1** [comment]
When desperately tried removing and repairing, myTouchHome (more talkative than EVE) stated that the error was a *value higher than maximum* 
I am just wondering which value that could be. Removing the thermostat immediately enables repairing after clearing homekit and persist.

**Comment by snowdd1** [comment]
Ok, found it. The TargetTemperature came immediately back with a value of 0, which is below the minimumValue of 10. 
Strange values however, it should have been 7 (Default Frost Protection value with my thermostat), but that HK wouldn't have liked either.

So THAT issue is solved. 

Did you ever tried to put two lamp services into one accessory - and succeeded?

**Comment by KhaosT** [comment]
@snowdd1 you can override the minimal value to 0 or -100 so that HomeKit will pair with the accessory.

**Comment by KhaosT** [comment]
@snowdd1 and multiple services works for me ... https://gist.github.com/KhaosT/371bfd8a4c86acd78d2d

**Comment by snowdd1** [comment]
> override the minimal value
Did that, works fine, plus did a check when values come in on minimum and maximum.

>and multiple services works 
ARE THERE naming conventions on the subtypes? I tried to put the displayName as subtype as that was already there and unique - I saw you used no-whitespace & lowercase in subtype
````javascript
var light1 = new Service.Lightbulb("Light 1","1light"); 
var light2 = new Service.Lightbulb("Light 2","2light"); 
````
I had tried with `new Service.Lightbulb("Light 1","Light 1")` - etc. , and failed.

**Comment by KhaosT** [comment]
@snowdd1 subtype can be any string (with/without space)... I'm not sure why your's failed. Does server produce any error when that happens?

**Comment by snowdd1** [comment]
No it just looks strange in EVE. I put back in the `var myService = new Service.LockMechanism(config.name,config.name);` etc. and try again, let's see how it looks in myTouchHome.

**Comment by luxus** [comment]
hello,
i am new to this homebridge stuff, if i understand this topic right it is about merging 2 lamps into one virtual device right?
after i playing around with homebridge this was the first thing i hoping for, because i have 2 lamps with 3 bulbs.
but the problem is i don't understand how to do it :D

**Comment by snowdd1** [comment]
frankly, I do not know what I messed up the last two times, but now it's working like a charm.
In EVE the service names appear neatly next two the accessory name. I had added a device "Office" with "Office lamp" and "Aisle lamp".

While I type HK or EVE gets confused in the background. My two lights first vanished, then reappeared as a duplicates of the latter light - Aisle lamp is there twice, and the Office lamp is gone. If I change the power state, always Aisle lamp is switched.

But in myTouchHome I still see two "Lightbulb"s without names, that control my Office lamp and my Aisle lamp separately. 

So I guess it's a bug in EVE after all. Let's wait for the updates for iOS9 then.
Good night :smile:
